Add a migration guide for k6 browser v0.52 (#1604)
grafana/k6-docs: docs/sources/v0.51.x/using-k6-browser/migrating-to-k6-v0-52.md

> ---

Author: This page will be published under the current release (`v0.51`). Once this page has been approved, I'll work on copying this page, and making the necessary amendments to it for the next release (`v0.52`).

If you are interested in the rational as to why we're making this change, have a look at the [v0.51 release notes](https://github.com/grafana/k6/releases/tag/v0.51.0).

## Migrating to async

To ensure your scripts work with the upcoming release of k6, you will need to ensure that all the synchronous (sync) APIs that are going to be migrated over to async are handled appropriately. In majority of cases you will only need to add `await` in front of the API call, like so:

From

{{< code >}}

```javascript
import { browser } from 'k6/experimental/browser';

export default async function () {
  const page = browser.newPage();
  // ...
}
```

{{< /code >}}

To

{{< code >}}

```javascript
import { browser } from 'k6/experimental/browser';

export default async function () {
  const page = await browser.newPage();
  // ...
}
```

{{< /code >}}

You will likely already have encountered async APIs when working with the browser module such as [page.click]({{< relref "../javascript-api/k6-experimental/browser/page/click/" >}}), so the use of `async` and `await` keywords will be familiar to you.

> The full list of APIs that will migrate over to be async can be found in this [github comment](https://github.com/grafana/xk6-browser/issues/428#issuecomment-1964020837). You will also find a screenshot of a comparison between the [fillform.js example](https://github.com/grafana/xk6-browser/blob/main/examples/fillform.js) in `v0.51` and `v0.52` in another [comment](https://github.com/grafana/xk6-browser/issues/428#issuecomment-2141634960) in the same github issue.

Reviewer: I would put that list and image here in this migration document for immediate visibility instead of linking it to a comment on a Github issue.

Author: Good spot, i was supposed to put a comment here asking for suggestions on this.

@heitortsergent what's the best way to add this very long list to the docs? Is it possible to have a collapsible list?

I'm assuming the image is straight forward. However it would be better if there is already a built in code comparison view?

Author: I've added the full list of APIs in 7b5f2d7c71003ce817231212de3a5cea997011c0

I've added a new screenshot comparing a simpler example script in 410edd5e7562f9099779073928be73dab12b7478

Author: I've found a way to collapse long lists.

### When can I migrate?

> Awaiting on something that’s not a thenable just returns that value. In another words, you can add the `await` keyword today to the browser module API calls in your scripts to future proof them. For example this works in k6 `v0.51` and it will work in k6 `v0.52`:

Reviewer: "It is safe to add the `await` keyword in front of the methods to be ready for the upcoming changes..."?

Author: ```suggestion
Awaiting on something that’s not a thenable just returns that value.  In another words, it is safe to add the `await` keyword in front of the browser module API methods in your existing scripts to be ready for the upcoming changes. For example this works in k6 `v0.51` and it will work in k6 `v0.52`:
```

WDYT?

Author: Amended in 5399033b58c311aeeff59dd9643aaf060c89bebc

{{< code >}}

```javascript
import { check } from 'k6';
import { browser } from 'k6/experimental/browser';

export const options = {
  scenarios: {
    ui: {
      executor: 'shared-iterations',
      options: {
        browser: {
          type: 'chromium',
        },
      },
    },
  },
  thresholds: {
    checks: ['rate==1.0'],
  },
};

export default async function () {
  const context = await browser.newContext();
  const page = await context.newPage();

  try {
    // Navigate to the website under test.
    await page.goto('https://test.k6.io/', { waitUntil: 'networkidle' });
    await Promise.all([
      page.waitForNavigation(),
      page.locator('a[href="/my_messages.php"]').click(),
    ]);

    // Check that the header reads "Unauthorized".
    const h2 = page.locator('h2');
    const headerOK = (await h2.textContent()) == 'Unauthorized';
    check(headerOK, { header: headerOK });

    // Take a screenshot of the page.
    await page.screenshot({ path: 'screenshot.png' });
  } finally {
    await page.close();
  }
}
```

{{< /code >}}

> ### Can I promise chain?

Reviewer: To make it clear in which version they can't do promise chaining.

```suggestion
### Can I promise chain in v0.51?
```

export default async function () {
  const page = browser.newPage();

  // ...

  const text = await page.$('#text1');
  await text.type('hello world');

  // ...
}
```

{{< /code >}}

## Working with k6 check

The k6 `check` API wasn't designed to work with async APIs. However there is a workaround. In your current scripts, you may use a `check` like so:

{{< code >}}

```javascript
import { browser } from 'k6/experimental/browser';
import { check } from 'k6';

export default async function () {
  const page = browser.newPage();

  // ...

  check(page, {
    header: (p) => p.locator('h2').textContent() == 'Welcome, admin!',
  });

  // ...
}
```

{{< /code >}}

Since the `locator.textContent` API is being migrated to async, and `check` doesn't work with async APIs, you will need to do something like this:

{{< code >}}

```javascript
import { browser } from 'k6/experimental/browser';
import { check } from 'k6';

export default async function () {
  const page = browser.newPage();

  // ...

  const h2 = page.locator('h2');
  const headerOK = (await h2.textContent()) == 'Welcome, admin!';
  check(headerOK, { header: headerOK });

  // ...
}
```

{{< /code >}}

## Groups

> A note on groups, they do not work with async APIs either, there is no workaround as of yet. Here's the [github issue](https://github.com/grafana/k6/issues/2728) that you can follow to keep up to date with relevant news on a groups API that works with async APIs.

Author: Is it worth mentioning this? It's likely that people have already tried this and realised that it doesn't work since a `group` might contain a `page.goto` or `page.click` API call, which are both async already.

Reviewer 2: My two cents, I think it's worth adding it here just to be explicit that this is a limitation. It could save someone time versus discovering it on their own.

# Are there any other upcoming major breaking changes?

> We think that this is likely the last large breaking change until the browser module is deemed to be stable enough to be a non-experimental module in k6. When the browser module is no longer experimental, we think it should only be a small change in the import path from `k6/experimental/browser` to `k6/browser`. We have an open [issue on github](https://github.com/grafana/xk6-browser/issues/1117) which you can view and track to see the progress of graduating the browser module out of experimental. For more information on graduating extension out of experimental, refer to the [extension graduation process]({{< relref "../extensions/explanations/extension-graduation" >}}).

Reviewer: I wouldn't link to the non-experimental issue just yet because most tasks in that issue have not been done yet, and it also explains that the module is not stable, etc. Although we're at an experimental stage, I wouldn't directly advertise it like so.

Author: Hmm, yep, I see what you are saying, and it's not particularly user friendly. I was trying to emphasis that this was seriously on the cards and people should be ready for another possible breaking change. At the same time, this could dissuade users from even trying to work with k6 browser if they saw this.

Author: Removed in 4978eebbffe68ccbb165f1978df3ccdd4402f825

# Where to get extra help

We will also be releasing the latest type definitions and updating the documentation when we release k6 `v0.52`. Refer to the [k6 IntelliSense]({{< relref "../set-up/configure-k6-intellisense" >}}) documentation on how to work with the type definition file in your project.

> For all the details, make sure to review the **upcoming** changelog for [k6 version 0.52](https://github.com/grafana/k6/releases/tag/v0.52.0). As always, ask in [the community forum](https://community.grafana.com/c/grafana-k6/k6-browser/79) if you need help!

Author: Should I remove the link to the missing release notes file for now?

Reviewer 2: Yeah, I wouldn't add a missing link here. I think we could change it to:

```suggestion
For all the details, make sure to review the _upcoming_ CHANGELOG for [k6 version 0.52](https://github.com/grafana/k6/releases/). You can also ask questions in [the community forum](https://community.grafana.com/c/grafana-k6/k6-browser/79) if you need help.
```

Author: Thanks! I've added this suggestion and also all the suggestions from https://github.com/grafana/k6-docs/pull/1606

{{< code >}}

```javascript
import { check } from 'k6';
import { browser } from 'k6/experimental/browser';

export const options = {
  scenarios: {
    ui: {
      executor: 'shared-iterations',
      options: {
        browser: {
          type: 'chromium',
        },
      },
    },
  },
  thresholds: {
    checks: ['rate==1.0'],
  },
};

export default async function () {
  const context = await browser.newContext();
  const page = await context.newPage();

  try {
    // Navigate to the website under test.
    await page.goto('https://test.k6.io/', { waitUntil: 'networkidle' });
    await Promise.all([
      page.waitForNavigation(),
      page.locator('a[href="/my_messages.php"]').click(),
    ]);

    // Check that the header reads "Unauthorized".
    const h2 = page.locator('h2');
    const headerOK = (await h2.textContent()) == 'Unauthorized';
    check(headerOK, { header: headerOK });

    // Take a screenshot of the page.
    await page.screenshot({ path: 'screenshot.png' });
  } finally {
    await page.close();
  }
}
```

{{< /code >}}

### Can I promise chain in v0.51?

> If you want to promise chain the newly migrated async APIs, You won't be able to do that until after the migration and release of k6 `v0.52`. For example, you can't do this yet:

Author: Thanks for the feedback, what about this: f739cec6ca104f2156029673ea813f83f330b946

> If you want to promise chain the newly migrated async APIs, You won't be able to do that until after the migration and release of k6 `v0.52`. For example, you can't do this yet:

Reviewer 3: Looks good! I would still prefer a `.` instead of `,` after the "No" though. 🙂

> If you want to promise chain the newly migrated async APIs, You won't be able to do that until after the migration and release of k6 `v0.52`. For example, you can't do this yet:

Author: Amended in ee78bc5293d885c7c393f2382bf9819df8d93e52

  const page = browser.newPage();

  // ...

  const text = await page.$('#text1');
  await text.type('hello world');

  // ...
}
```

{{< /code >}}

## Working with k6 check

The k6 `check` API wasn't designed to work with async APIs. However there is a workaround. In your current scripts, you may have used a `check` like so:

{{< code >}}

```javascript
import { browser } from 'k6/experimental/browser';
import { check } from 'k6';

// options block

export default async function () {
  const page = browser.newPage();

  // ...

  check(page, {
    header: (p) => p.locator('h2').textContent() == 'Welcome, admin!',
  });

  // ...
}
```

{{< /code >}}

Since the `locator.textContent` API is being migrated to async, and `check` doesn't work with async APIs, you will need to do something like this:

{{< code >}}

```javascript
import { browser } from 'k6/experimental/browser';
import { check } from 'k6';

// options block

export default async function () {
  const page = browser.newPage();

  // ...

  const h2 = page.locator('h2');
  const headerOK = (await h2.textContent()) == 'Welcome, admin!';
  check(headerOK, { header: headerOK });

  // ...
}
```

> {{< /code >}}

Author: I like it. The issue is with the linter, it complains that check, page etc aren't defined in the example. I've amended your example a tiny bit to avoid the linter issue. Updated in 38cf3a175ddd13f9dd7bacbdf4a798908fc945ba.

@heitortsergent is there a way to bypass the javascript linter?

> {{< /code >}}

Author: I found a way to skip the linter.
